pkg/controller/privatelink/conditions/conditions_test.go

```diff
@@ -0,0 +1,358 @@
+	testReason            = "test-reason"
+	testMessage           = "test-message"
+)
+
+func Test_SetErrCondition(t *testing.T) {
+	scheme := scheme.GetScheme()
+	cdBuilder := testcd.FullBuilder(testNameSpace, testClusterDeployment, scheme)
+
+	cases := []struct {
+		name     string
+		cd       *hivev1.ClusterDeployment
+		cdExists bool
+
+		expectConditions []hivev1.ClusterDeploymentCondition
+		expectError      string
+		expectLogs       []string
+	}{{ // There should be an error when the clusterdeployment is not found
+		name:        "clusterdeployment not found",
+		cd:          cdBuilder.Build(),
+		expectError: "failed to get cluster deployment: clusterdeployments.hive.openshift.io \"" + testClusterDeployment + "\" not found",
+	}, { // There should be a log message when the failed condition changed to true
+		name: "failed condition changed to true",
+		cd: cdBuilder.Build(
+			testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+				Type:    hivev1.PrivateLinkFailedClusterDeploymentCondition,
+				Status:  corev1.ConditionFalse,
+				Reason:  testReason,
+				Message: testMessage,
+			}),
+			testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+				Type:    hivev1.PrivateLinkReadyClusterDeploymentCondition,
+				Status:  corev1.ConditionFalse,
+				Reason:  testReason,
+				Message: testMessage,
+			}),
+		),
+		cdExists: true,
+		expectConditions: []hivev1.ClusterDeploymentCondition{{
+			Status:  corev1.ConditionTrue,
+			Type:    hivev1.PrivateLinkFailedClusterDeploymentCondition,
+			Reason:  testReason,
+			Message: testMessage,
+		}, {
+			Status:  corev1.ConditionFalse,
+			Type:    hivev1.PrivateLinkReadyClusterDeploymentCondition,
+			Reason:  testReason,
+			Message: testMessage,
+		}},
+		expectLogs: []string{"setting PrivateLinkFailedClusterDeploymentCondition to true"},
+	}, { // There should be a log message when the ready condition changed to false
+		name: "ready condition changed to false",
+		cd: cdBuilder.Build(
+			testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+				Type:    hivev1.PrivateLinkFailedClusterDeploymentCondition,
+				Status:  corev1.ConditionTrue,
+				Reason:  testReason,
+				Message: testMessage,
+			}),
+			testcd.WithCondition(hivev1.ClusterDeploymentCondition{
+				Type:    hivev1.PrivateLinkReadyClusterDeploymentCondition,
+				Status:  corev1.ConditionTrue,
+				Reason:  testReason,
+				Message: testMessage,
+			}),
+		),
+		expectConditions: []hivev1.ClusterDeploymentCondition{{
+			Status:  corev1.ConditionTrue,
+			Type:    hivev1.PrivateLinkFailedClusterDeploymentCondition,
+			Reason:  testReason,
+			Message: testMessage,
+		}, {
+			Status:  corev1.ConditionFalse,
+			Type:    hivev1.PrivateLinkReadyClusterDeploymentCondition,
+			Reason:  testReason,
+			Message: testMessage,
+		}},
+		cdExists:   true,
+		expectLogs: []string{"setting PrivateLinkFailedClusterDeploymentCondition to true"},
+	}, { // There should be no errors when no conditions change
```

Comment:
I'm taking advantage of the logged messages to ensure we are reaching the specified cases as opposed to similar code paths that may return the same value(s). This helps avoid the false positives for the test case where the same result was returned from another unexpected code path. As an example, testing "two case" in the following while ensuring "one case" was not the reason "result" was returned:

```
func something(arg string) string {
    if arg == "one case" {
        log("one case reached)
        return "result"
   }
   if arg == "two case" {
       log("two case reached)
       return "result"
   }
   return ""
}
```

Reply:
I'm okay with that. So the question remains: we don't have a negative assertion to prove that the path with the log was *not* taken. Should we?

Reply:
I mean, we could. It would take an additional test variable, etc. I was generalizing in my prior comment about using logging to help test. In this function, it may be diminishing return when combined with the expectedStatus (which is something I added later).

Reply:
:shrug:

Reply:
I think I'm good as is. We are testing the expectedConditions which identifies which path the code took as they are not changed from the preconfigured conditions. Using logging to do this would be redundant and unnecessarily complicate the code.

Reply:
I've decided to [rewrite](https://github.com/openshift/hive/compare/e40350258a0584e1d5107e463ed191f1a1f63aa8..2b6c54a889316e403abdfdbf9a8ee0477e01b2fe) these to be geared to testing the resulting conditions.